### ISSUE TITLE
SyncDictionaries clash with Sirenix.OdinInspector

### ISSUE THREAD

**Issue by 0risin** [FirstGearGames/FishNet#9]
Fishnet: 1.1.4
Unity: 2021.2.7f1
**SyncDictionaries clash with Sirenix.OdinInspector**

When a 2nd player joins, the SyncDictionary in a PlayerComponent inexplicably throws a null pointer exception, key and value seem to be filled but the NetworkBehaviour is null. I do not know what is null in this constellation.

```cs
NullReferenceException  : Object reference not set to an instance of an object
FishNet.Object.Synchronizing.SyncIDictionary`2[TKey,TValue].AddOperation (FishNet.Object.Synchronizing.SyncDictionaryOperation operation, TKey key, TValue value) (at Assets/FishNet/Runtime/Object/Synchronizing/SyncDictionary.cs:153)
FishNet.Object.Synchronizing.SyncIDictionary`2[TKey,TValue].Add (TKey key, TValue value, System.Boolean asServer) (at Assets/FishNet/Runtime/Object/Synchronizing/SyncDictionary.cs:370)
FishNet.Object.Synchronizing.SyncIDictionary`2[TKey,TValue].Add (TKey key, TValue value) (at Assets/FishNet/Runtime/Object/Synchronizing/SyncDictionary.cs:364)
FishNet.Object.Synchronizing.SyncIDictionary`2[TKey,TValue].Add (System.Collections.Generic.KeyValuePair`2[TKey,TValue] item) (at Assets/FishNet/Runtime/Object/Synchronizing/SyncDictionary.cs:355)
Sirenix.Serialization.GenericCollectionFormatter`2[TCollection,TElement].DeserializeImplementation (TCollection& value, Sirenix.Serialization.IDataReader reader) (at Y:/Repositories/sirenix-development-framework/OdinSerializer/OdinSerializer/Core/Formatters/GenericCollectionFormatter.cs:132)
UnityEngine.Debug:LogException(Exception)
Sirenix.Serialization.CustomLogger:LogException(Exception) (at Y:/Repositories/sirenix-development-framework/Sirenix Solution/Sirenix.Serialization.Config/CustomLogger.cs:67)
Sirenix.Serialization.DebugContext:LogException(Exception) (at Y:/Repositories/sirenix-development-framework/OdinSerializer/OdinSerializer/Core/Misc/SerializationConfig.cs:252)
Sirenix.Serialization.GenericCollectionFormatter`2:DeserializeImplementation(SyncDictionary`2&, IDataReader) (at Y:/Repositories/sirenix-development-framework/OdinSerializer/OdinSerializer/Core/Formatters/GenericCollectionFormatter.cs:136)
Sirenix.Serialization.BaseFormatter`1:Deserialize(IDataReader) (at Y:/Repositories/sirenix-development-framework/OdinSerializer/OdinSerializer/Core/Formatters/BaseFormatter.cs:222)
Sirenix.Serialization.ComplexTypeSerializer`1:ReadValue(IDataReader) (at Y:/Repositories/sirenix-development-framework/OdinSerializer/OdinSerializer/Core/Serializers/ComplexTypeSerializer.cs:366)
Sirenix.Serialization.Serializer`1:ReadValueWeak(IDataReader) (at Y:/Repositories/sirenix-development-framework/OdinSerializer/OdinSerializer/Core/Serializers/Serializer.cs:261)
Sirenix.Serialization.UnitySerializationUtility:DeserializeUnityObject(Object, IDataReader) (at Y:/Repositories/sirenix-development-framework/OdinSerializer/OdinSerializer/Unity Integration/UnitySerializationUtility.cs:1881)
Sirenix.Serialization.UnitySerializationUtility:DeserializeUnityObject(Object, Byte[]&, List`1&, DataFormat, DeserializationContext) (at Y:/Repositories/sirenix-development-framework/OdinSerializer/OdinSerializer/Unity Integration/UnitySerializationUtility.cs:1730)
Sirenix.Serialization.UnitySerializationUtility:DeserializeUnityObject(Object, SerializationData&, DeserializationContext, Boolean, List`1) (at Y:/Repositories/sirenix-development-framework/OdinSerializer/OdinSerializer/Unity Integration/UnitySerializationUtility.cs:1415)
```

I would assume that this is the SyncDict from the 1st Player Object that joined.

The code that throws the null pointer

```cs

        public void Add(TKey key, TValue value)
        {
            Add(key, value, true);
        }
        private void Add(TKey key, TValue value, bool asServer)
        {
            Collection.Add(key, value);
            if (asServer)
                AddOperation(SyncDictionaryOperation.Add, key, value);
        }
```

The dictionary key and value are not null as tested in the Debugger

**Comment by FirstGearGames** [comment]
Try removing or commenting out odin properties and see if problem persist.

**Comment by 0risin** [comment]
I will provide information as soon as I test 2 Clients 1 Server constellation again.

**Comment by NickMercer** [comment]
I have this issue as well. For me it happens on an object owned by the host (haven't tested clients because this happens on host start and I can't get that far.)

Same as above, the AddOperation method in the SyncDictionary class has a correct key and value but the base.NetworkBehaviour is null. 

I do have Odin inspector in my project, and I removed any reference to it from the NetworkBehaviour with the SyncDictionary and the class I'm serializing as the Value in the dictionary. Not even a using statement.

Looking through the debugger, the SyncDictionary's NetworkManager is also null. It looks like something is causing the SyncDictionary to not get initialized. Everything I can see is at default values.

Update while writing this: I ended up fixing this when I noticed that SyncDictionary didn't have [SyncObject] above it. I didn't get any errors or warnings in the console when this was the case, but putting SyncObject over it made the "Must be readonly" error appear, which I fixed, and then the addition worked fine. 

Since I'm moving a project over from Mirror, I didn't have the SyncObject attribute, and I think this error probably just comes down to figuring out a way to surface the missing attribute to the user, if it's necessary for the object to work correctly.
